### PR TITLE
[ENH][DOC] local join count and LOSH statistics

### PR DESCRIPTION
**Pull request by jeffcsauer**
**Purpose of PR**

This PR addresses bullet point 1, 5, and 6 of #61. Specifically, this PR adds four new functions to estimate:
- Local spatial heteroskedasticity (function name: `losh`)
- Local join counts - univariate (function name: `Local_Join_Count`)
- Local join counts - bivariate (function name: `Local_Join_Count_BV`)
- Local join counts - multivariate (function name: `Local_Join_Count_MV`)

Each function is written in the form of a scikit-learn style estimator. PEP8 formatting has been applied to all of the functions, although a handful of lines are left long due to readability.  

**Documentation**

Two notebooks are included in the PR. Each notebook briefly reviews the core math of the statistic and explains how a user might deploy the statistic in a geospatial workflow. The included examples match those found in [`R` `spdep::LOSH` and `spdep::LOSH.cs`](https://www.google.com/search?client=firefox-b-1-d&q=losh.cs) as well as the [GeoDa Local Join Counts tutorial](https://www.google.com/search?client=firefox-b-1-d&lei=ytAVX7L4EYKztAb23K9A&q=local%20indicators%20of%20spatial%20association&ved=2ahUKEwiyhYuJqtzqAhWCGc0KHXbuCwgQsKwBKAB6BAgPEAE&biw=1920&bih=938).  

All functions include docstrings and doctests that use built-in PySAL example datasets.

**Tests**

I have added in four .py test files that execute successfully [on my fork](https://github.com/jeffcsauer/esda). These new tests follow the lead of the existing tests, specifically `test_moran.py` for the LOSH function and `test_join_counts.py` for the local join count functions.

**Notes for PR consideration**

- Local join count functions are currently using the default `n_jobs=1` to avoid issues fixed in potential [merges](#diff-1fd239570b4623d6ce23f0e268495633L510). Users may enter values other than 1. 
- One doctest dataset, `commpop`, appears to have a [dead link](https://geodacenter.github.io/data-and-lab/) at the moment. Having trouble loading this as a `libpysal.example`. Once this dataset is working the doctest can be updated. Currently hosting it on a personal Github link.

### CODE REVIEW

File: esda/local_join_count.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+        """
+
+        self.connectivity = connectivity
+        self.permutations = permutations
+        self.n_jobs = n_jobs
+        self.keep_simulations = keep_simulations
+        self.seed = seed
+
+    def fit(self, y, n_jobs=1, permutations=999):
```
Comment (ljwolf) [inline esda/local_join_count.py:63]: If you define permutations above, then you should probably not set the default here again! 

Also `sklearn` estimators generally put these kinds of options in the init, not fit. And, all `sklearn` estimators, even the unsupervised ones, [take both an X and a y](https://scikit-learn.org/stable/developers/develop.html#fitting)... we should follow those conventions, I think.

File: esda/local_join_count_bv.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                           
+        """
+
+        self.connectivity = connectivity
+        self.permutations = permutations
+        self.n_jobs = n_jobs
+        self.keep_simulations = keep_simulations
+        self.seed = seed
+
+    def fit(self, x, z, case="CLC", n_jobs=1, permutations=999):
```
Comment (ljwolf) [inline esda/local_join_count_bv.py:57]: Again, generally, sklearn calls these `X` and `y`. Can we use that here?

File: esda/losh.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
+import numpy as np
+import warnings
+from scipy import sparse
+from scipy import stats
+from sklearn.base import BaseEstimator
+import libpysal as lp
+
+
+class losh(BaseEstimator):
```
Comment (ljwolf) [inline esda/losh.py:9]: This should be `LOSH` keeping with style elsewhere.

File: esda/losh.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                           Variance of Hi.
+        pval             : numpy array
+                           P-values for inference based on either
+                           "chi-square" or "permutation" methods.
+        """
+
+        self.connectivity = connectivity
+        self.inference = inference
+
+    def fit(self, y, a=2):
```
Comment (ljwolf) [inline esda/losh.py:43]: `a=2` should be in the init, and again on `X` and `y` as the arguments here.

**Review by ljwolf — CHANGES_REQUESTED** [review]
👏 this is amazing @jeffcsauer! well done! extremely minor API changes that mesh with the sklearn developer guide, but jeez, these are awesome!

File: esda/losh.py (hunk 3 above)
Comment (jeffcsauer) [reply esda/losh.py]: Changed to `LOSH` in https://github.com/pysal/esda/pull/139/commits/3b47da838127a6e9677158ac9d11a2b321a8bef9

File: esda/losh.py (hunk 4 above)
Comment (jeffcsauer) [reply esda/losh.py]: Moved `a=2` to init, and given that losh is univariate changed `y` to `x` in `fit(...)`, can you confirm on https://github.com/pysal/esda/pull/139/commits/3b47da838127a6e9677158ac9d11a2b321a8bef9?

File: esda/local_join_count.py (hunk 1 above)
Comment (jeffcsauer) [reply esda/local_join_count.py]: Fixed permutations such that it is part of `init(...)` and removed constant from top of file. Applied to all LJC functions. See https://github.com/pysal/esda/pull/139/commits/2eb0d37d059af4596a2ffd0e103820a7ae0dcd74.

File: esda/local_join_count_bv.py (hunk 2 above)
Comment (jeffcsauer) [reply esda/local_join_count_bv.py]: Changed for univariate and bivariate statistics. For multivariate statistics, kept input array as `variables`. Is this okay? Or should we change to `x`? Thought that `variables` is more explicit/clear compared to `x`, but totally open to changing it to be consistent! See https://github.com/pysal/esda/pull/139/commits/2eb0d37d059af4596a2ffd0e103820a7ae0dcd74.

**Comment by ljwolf** [comment]
This has been superseded by #145

**Pull request closed without merging**